SpatialPyramidPooling.lua

```diff
@@ -1,61 +1,38 @@
-local SpatialPyramidPooling, parent = torch.class('inn.SpatialPyramidPooling', 'nn.Module')
+local SpatialPyramidPooling, parent = torch.class('inn.SpatialPyramidPooling', 'nn.Concat')
 
-function SpatialPyramidPooling:__init(pyr)
-  parent.__init(self)
-  
-  self.pyr = pyr
-  
-  self.dimd = 1
-  self.module = nn.Concat(self.dimd)
-  for i=1,#self.pyr do
-    local t = nn.Sequential()
-    t:add(nn.SpatialAdaptiveMaxPooling(self.pyr[i][1],self.pyr[i][2]))
-    t:add(nn.View(-1):setNumInputDims(3))
-    self.module:add(t)
-  end
-  
-  self.output = torch.Tensor()
-  
+-- allows nn.SpatialPyramidPooling({4,4},{3,3}) or {{4,4},{3,3}}
+function SpatialPyramidPooling:__init(...)
+   parent.__init(self, 1)
+   local pyr = {...}
+   self.pyr = torch.type(pyr[1][1]) == 'table' and pyr[1] or pyr
+   self.dimd = 1
+   for k,v in ipairs(self.pyr) do
+      parent.add(self, nn.Sequential()
+         :add(nn.SpatialAdaptiveMaxPooling(v[1], v[2]))
+         :add(nn.View(-1):setNumInputDims(3))
+         :add(nn.Contiguous())
+      )
+   end
 end
 
-function SpatialPyramidPooling:updateOutput(input)
-  
-  local dimd = 1
-  if input:nDimension() == 4 then 
-    dimd = dimd + 1
-  end
+function dimensionSet(self, input)
+   local dimd = 1
+   if input:nDimension() == 4 then 
+      dimd = dimd + 1
+   end
 
-  if self.dimd ~= dimd then
-    self.dimd = dimd
-    self.module.dimension = dimd
-  end
-  
-  assert(input:type()==self.output:type(),'Wrong input type!')
-  
-  self.output = self.module:updateOutput(input)
-  return self.output
+   if self.dimd ~= dimd then
+      self.dimd = dimd
```

Comment:
I think the meaning of `dimd` is redundant now, and can be replaced by `dimension`, right ?